solr/core/src/java/org/apache/solr/filestore/NodeFileStore.java

```diff
@@ -0,0 +1,211 @@
+  //  multiple distinct APIs
+  @Override
+  @PermissionName(FILESTORE_READ_PERM)
+  public SolrJerseyResponse getFile(String path, Boolean sync, String getFrom, Boolean meta) {
+    final var response = instantiateJerseyResponse(SolrJerseyResponse.class);
+
+    if (Boolean.TRUE.equals(sync)) {
+      try {
+        fileStore.syncToAllNodes(path);
+        return response;
+      } catch (IOException e) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Error getting file ", e);
+      }
+    }
+
+    final var pathCopy = path;
+    if (getFrom != null) {
+      coreContainer
+          .getUpdateShardHandler()
+          .getUpdateExecutor()
+          .submit(
+              () -> {
+                log.debug("Downloading file {}", pathCopy);
+                try {
+                  fileStore.fetch(pathCopy, getFrom);
+                } catch (Exception e) {
+                  log.error("Failed to download file: {}", pathCopy, e);
+                }
+                log.info("downloaded file: {}", pathCopy);
+              });
+      return response;
+    }
+
+    if (path == null) {
+      path = "";
+    }
+    FileStore.FileType typ = fileStore.getType(path, false);
+    if (typ == FileStore.FileType.NOFILE) {
+      final var fileMissingResponse =
+          instantiateJerseyResponse(FileStoreDirectoryListingResponse.class);
+      fileMissingResponse.files = Collections.singletonMap(path, null);
+      return fileMissingResponse;
+    }
+    if (typ == FileStore.FileType.DIRECTORY) {
+      final var directoryListingResponse =
+          instantiateJerseyResponse(FileStoreDirectoryListingResponse.class);
+      final var directoryContents =
+          fileStore.list(path, null).stream()
+              .map(details -> convertToResponse(details))
+              .collect(Collectors.toList());
+      directoryListingResponse.files = Collections.singletonMap(path, directoryContents);
+      return directoryListingResponse;
+    }
+    if (Boolean.TRUE.equals(meta)) {
+      if (typ == FileStore.FileType.FILE) {
+        int idx = path.lastIndexOf('/');
+        String fileName = path.substring(idx + 1);
+        String parentPath = path.substring(0, path.lastIndexOf('/'));
+        List<FileStore.FileDetails> l = fileStore.list(parentPath, s -> s.equals(fileName));
+
+        final var fileMetaResponse =
+            instantiateJerseyResponse(FileStoreDirectoryListingResponse.class);
+        fileMetaResponse.files =
+            Collections.singletonMap(path, l.isEmpty() ? null : convertToResponse(l.get(0)));
+        return fileMetaResponse;
+      }
+    } else { // User wants to get the "raw" file
+      // TODO Should we be trying to json-ify otherwise "raw" files in this way?  It seems like a
+      // pretty sketchy idea, esp. for code with very little test coverage.  Consider removing
+      final var pathCopy2 = path;
```

Comment:
pathCopy2???

Reply:
This is something that carried over from the existing code, and I'm leery of getting sucked into larger improvements to that code. There's a lot that could be done there.
But I can change it since it's a small modification.